### PR TITLE
msglist: Set unreads.oldUnreadsMissing to false on batched mark-all-as-read

### PR DESCRIPTION
**Pull request by chrisbobbe**

### CODE REVIEW

File: lib/widgets/message_list.dart [hunk 1]
```diff
@@ -366,14 +366,24 @@ class MarkAsReadWidget extends StatelessWidget {
...
+      //
+      // [Unreads] is supposed to do this naturally, in
+      // handleUpdateMessageFlagsEvent. That's based on the documented behavior
+      // that a completed mark-all-as-read task gets signaled by `all: true` on
+      // the [UpdateMessageFlagsAddEvent]. But when using this new batch-
+      // processing protocol, it seems clients have to track progress and
+      // success themselves; the event from the last batch is unmarked.
+      // Discussion:
+      //   <https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/flutter.3A.20Mark-as-read/near/1680275>
+      PerAccountStoreWidget.of(context).unreads.oldUnreadsMissing = false;
```
Comment (sirpengi) [inline lib/widgets/message_list.dart:385]: `notifyListeners` isn't being called after mutating the `unreads` model here. I think there's an argument for not doing that here (as we aren't using this flag in the UI) but leaving a note to that effect might be good?

Comment (chrisbobbe) [reply lib/widgets/message_list.dart:385]: Oh right, thanks for catching that! I think it'd be good to call `notifyListeners`. We might in the future use this flag in the UI; here's from the doc on `old_unreads_missing` in the `/register` response:

> […] When true, we recommend that clients display a warning, as they are likely to produce erroneous results until reloaded with the user having fewer than MAX_UNREAD_MESSAGES unread messages.

I'll send a revision with one way of making that happen.

**Comment by chrisbobbe** [comment]
Thanks for the review! Revision pushed.

File: test/widgets/message_list_test.dart [hunk 2]
```diff
@@ -608,6 +626,19 @@ void main() {
           narrow: narrow, messages: [message], unreadMsgs: unreadMsgs);
         check(isMarkAsReadButtonVisible(tester)).isTrue();
 
+        int unreadsNotifiedCount = 0;
+        final unreadsModel = store.unreads;
```
Comment (gnprice) [inline test/widgets/message_list_test.dart:630]: nit: just inline this, it's more transparent and not materially longer

File: test/widgets/message_list_test.dart [hunk 3]
```diff
@@ -608,6 +626,19 @@ void main() {
...
+        unreadsModel.addListener(() {
+          unreadsNotifiedCount++;
+        });
+        // Might as well test with oldUnreadsMissing: true.
+        //
+        // We didn't fill the model with 50k unreads, so this is questionably
+        // realistic… but the 50k cap isn't actually API-guaranteed, and this is
+        // plausibly realistic for a hypothetical server that decides based on
+        // message age rather than the 50k cap.
+        unreadsModel.oldUnreadsMissing = true;
```
Comment (gnprice) [inline test/widgets/message_list_test.dart:640]: I think this "might as well" reasoning works so long as this wrinkle isn't adding so much to read (in code and comments) that it comes to take over the test case.

One simplification is that I think this test is just as effective without the `unreadsNotifiedCount` part, just checking that `oldUnreadsMissing` is still true. So:

```suggestion
        // Might as well test with oldUnreadsMissing: true.
        store.unreads.oldUnreadsMissing = true;
```

File: test/widgets/message_list_test.dart [hunk 4]
```diff
@@ -618,6 +649,14 @@ void main() {
           ..bodyFields.deepEquals({});
 
         await tester.pumpAndSettle(); // process pending timers
+
+        // Didn't redundantly call [Unreads.handleAllMessagesReadSuccess], which
+        // is meant only for the modern protocol and would be redundant in the
+        // legacy protocol. (In the legacy protocol, we set oldUnreadsMissing to
+        // false when an event comes, not when the mark-as-read request
+        // succeeds.)
```
Comment (gnprice) [inline test/widgets/message_list_test.dart:657]: In the same vein:
```suggestion
        // Check that [Unreads.handleAllMessagesReadSuccess] wasn't called;
        // in the legacy protocol, that'd be redundant with the mark-read event.
```

File: test/widgets/message_list_test.dart [hunk 5]
```diff
@@ -570,6 +571,23 @@ void main() {
...
       });
 
+      testWidgets('markNarrowAsRead on mark-all-as-read when Unreads.oldUnreadsMissing: true', (tester) async {
+        const narrow = AllMessagesNarrow();
+        await setupMessageListPage(tester,
+          narrow: narrow, messages: [message], unreadMsgs: unreadMsgs);
+        check(isMarkAsReadButtonVisible(tester)).isTrue();
+        store.unreads.oldUnreadsMissing = true;
+
+        final connection = store.connection as FakeApiConnection;
```
Comment (gnprice) [inline test/widgets/message_list_test.dart:581]: nit: there's an outer `connection` that this is equivalent to, in the same way as `store`:
```suggestion
```

(I see there's a bunch of existing cases of similar redundancy in this file. I'll send a quick cleanup for those.)

**Review by gnprice — COMMENTED** [review]
Thanks to you both! Sorry for the delayed review. This looks good, with a few nits below.

**Comment by chrisbobbe** [comment]
Thanks for the review! Revision pushed.

File: test/widgets/message_list_test.dart (hunk 5 above)
Comment (gnprice) [reply test/widgets/message_list_test.dart:766]: bump

**Review by gnprice — COMMENTED** [review]
Thanks! One nit still open, though.

**Comment by chrisbobbe** [comment]
Ah indeed; thanks for catching that. Revision pushed!

**Comment by gnprice** [comment]
Thanks! Looks good; merging.

**Pull request merged**